train.py

```diff
@@ -69,46 +70,88 @@ def get_args():
+    total_length = sum(lengths)
+    attention_mask[:total_length] = 1
+    return attention_mask
+
+
+def preprocess_function(examples, args, is_train):
     inputs = examples["input"]
     targets = examples["output"]
-    if args.data_packing:
-        estimate_maximum_char = 4.0 * args.max_input_length
-        packed_inputs, context_len = [], 0
-        temp_example = ""
-        for inp_example, tgt_example in zip(inputs, targets):
-            # cannot fit in the current context
-            if context_len + len(inp_example) + len(tgt_example) > estimate_maximum_char:
-                # push the current context
-                packed_inputs.append(temp_example)
-                # refresh the context
-                temp_example = ""
-                context_len = 0
-            temp_example += inp_example + tgt_example + tokenizer.eos_token
-            context_len += len(inp_example) + len(tgt_example) + 1
-        if temp_example:
-            packed_inputs.append(temp_example)
-        model_inputs = packed_inputs
+    model_inputs = [inp + tar + tokenizer.eos_token for inp, tar in zip(inputs, targets)]
+
+    # do not do packing for fair comparison with no packing
+    if args.data_packing and is_train:
+        model_inputs = tokenizer(model_inputs,
+                                 max_length=args.max_input_length,
+                                 # no padding to calculate the real length
+                                 padding=False,
+                                 truncation=True)
+        # packing in the input_ids level
+        max_window = args.max_input_length
+        # packed_length maintains a list, each of which is a list of all packed sequence lengths in this sequence
+        packed_lengths, packed_ids,  = [], []
+        temp_lengths, temp_ids = [], []
+        for input_ids in model_inputs["input_ids"]:
+            if sum(temp_lengths) + len(input_ids) >= max_window:
+                packed_lengths.append(temp_lengths.copy())
+                packed_ids.append(temp_ids.copy())
+                # refresh for the next packed sequence
+                temp_lengths.clear()
+                temp_ids.clear()
+            temp_ids.extend(input_ids)
+            temp_lengths.append(len(input_ids))
+        # add the last one
+        if temp_lengths:
+            packed_lengths.append(temp_lengths.copy())
+            packed_ids.append(temp_ids.copy())
+        # pad the packed_ids into max_window
+        padded_input_ids = np.zeros((len(packed_ids), max_window), dtype=np.int64)
+        padded_attention_masks = np.zeros((len(packed_ids), max_window), dtype=np.int64)
+        for i, example_input_ids in enumerate(packed_ids):
+            padded_input_ids[i, :len(example_input_ids)] = example_input_ids
+            # obtain the attention mask
+            attention_mask = create_flatten_attention_mask(packed_lengths[i], args.max_input_length)
+            padded_attention_masks[i] = attention_mask
```

Comment:
```suggestion
            padded_attention_masks[i, :sum(packed_lengths[i])] += 1
```

If we make `packed_lengths` just store the total length instead, we can also save the sum